RHDEVDOCS-4967-update-nodeselector-example-for-moving-monitoring-components (#58755)
openshift/openshift-docs: modules/monitoring-using-node-selectors-to-move-monitoring-components.adoc

// Module included in the following assemblies:
//
// * monitoring/configuring-the-monitoring-stack.adoc

:_content-type: CONCEPT
[id="using-node-selectors-to-move-monitoring-components_{context}"]
= Using node selectors to move monitoring components

By using the `nodeSelector` constraint with node labels, you can move any of the monitoring stack components to specific nodes.
Doing so enables you to control the placement and distribution of the monitoring components across the cluster.

By specifying the nodes on which the components will be deployed, you can optimize system resource use, improve performance, and segregate workloads based on specific requirements or policies.

> [IMPORTANT]

Author: Added a new section of conceptual information explaining potential conflicts with existing constraints and anti-affinity rules. PTAL.

> [IMPORTANT]

Reviewer: done

====
To maintain high availability, some monitoring components are deployed by using multiple pods across different nodes in a cluster.
For example, hard anti-affinity constraints are set on Prometheus, Thanos Querier, and Alertmanager pods.
If you set a node selector that only matches one node, the scheduler will not be able to satisfy all other constraints and will not schedule a pod onto a node.

Therefore, when moving monitoring components to labeled nodes, ensure that enough matching nodes are available to maintain resilience for the component.

> If only one label is specified, ensure that enough nodes contain that label to distribute all of the pods for the component across separate nodes.

Author: Rewrote this entire section and added a paragraph to note that constraints are compounded when the pod scheduler is determining nodes on which to place the pods. PTAL. Thanks!

> If only one label is specified, ensure that enough nodes contain that label to distribute all of the pods for the component across separate nodes.

Reviewer: done

// Module included in the following assemblies:
//
// * monitoring/configuring-the-monitoring-stack.adoc

:_content-type: CONCEPT
[id="using-node-selectors-to-move-monitoring-components_{context}"]
= Using node selectors to move monitoring components

By using the `nodeSelector` constraint with labeled nodes, you can move any of the monitoring stack components to specific nodes.
By doing so, you can control the placement and distribution of the monitoring components across a cluster.

By controlling placement and distribution of monitoring components, you can optimize system resource use, improve performance, and segregate workloads based on specific requirements or policies.

[id="how-node-selectors-work-with-other-constraints_{context}"]
== How node selectors work with other constraints

If you move monitoring components by using node selector constraints, be aware that other constraints to control pod scheduling can exist for a cluster:

* Topology spread constraints might be in place to control pod placement.

> * Hard anti-affinity rules are in place for Prometheus, Thanos Querier, and Alertmanager components to ensure that multiple pods for these monitoring components are always spread across different nodes and are therefore always highly available.

Reviewer 2: Hard anti-affinity rules are also applied for prometheus-adapter/prometheus-operator-admission-webhook, and prometheus-operator-admission-webhook is added since 4.11, maybe we can change
`Hard anti-affinity rules are in place for Prometheus, Thanos Querier, and Alertmanager components`
to
`Hard anti-affinity rules are in place for Prometheus, Thanos Querier, Alertmanager and etc. components`

```
# oc -n openshift-monitoring get deploy prometheus-adapter -oyaml
...
    spec:
      affinity:
        podAntiAffinity:
          requiredDuringSchedulingIgnoredDuringExecution:
          - labelSelector:
              matchLabels:
                app.kubernetes.io/component: metrics-adapter
                app.kubernetes.io/name: prometheus-adapter
                app.kubernetes.io/part-of: openshift-monitoring
            namespaces:
            - openshift-monitoring
            topologyKey: kubernetes.io/hostname
...

# oc -n openshift-monitoring get deploy prometheus-operator-admission-webhook -oyaml
...
    spec:
      affinity:
        podAntiAffinity:
          requiredDuringSchedulingIgnoredDuringExecution:
          - labelSelector:
              matchLabels:
                app.kubernetes.io/name: prometheus-operator-admission-webhook
                app.kubernetes.io/part-of: openshift-monitoring
            namespaces:
            - openshift-monitoring
            topologyKey: kubernetes.io/hostname
...
```

Author: Changed wording to: `Hard anti-affinity rules are in place for Prometheus, Thanos Querier, Alertmanager, and other monitoring components`

When scheduling pods onto nodes, the pod scheduler tries to satisfy all existing constraints when determining pod placement.
That is, all constraints compound when the pod scheduler determines which pods will be placed on which nodes.

> Therefore, if you configure a node selector constraint so that existing constraints cannot all be satisfied, the pod scheduler will not be able to match all constraints and will not schedule a pod for placement onto a node.

Reviewer 2: ```suggestion
Therefore, if you configure a node selector constraint but existing constraints cannot all be satisfied, the pod scheduler will not be able to match all constraints and will not schedule a pod for placement onto a node.
```

> To maintain resilience and high availability for monitoring components, ensure that enough nodes are available that match all constraints when you configure a node selector constraint to move a component.

Reviewer 2: ```suggestion
To maintain resilience and high availability for monitoring components, ensure that enough nodes are available and match all constraints when you configure a node selector constraint to move a component.
```

Author: Changed

// Module included in the following assemblies:
//
// * monitoring/configuring-the-monitoring-stack.adoc

:_content-type: CONCEPT
[id="using-node-selectors-to-move-monitoring-components_{context}"]
= Using node selectors to move monitoring components

By using the `nodeSelector` constraint with labeled nodes, you can move any of the monitoring stack components to specific nodes.
By doing so, you can control the placement and distribution of the monitoring components across a cluster.

By controlling placement and distribution of monitoring components, you can optimize system resource use, improve performance, and segregate workloads based on specific requirements or policies.

[id="how-node-selectors-work-with-other-constraints_{context}"]
== How node selectors work with other constraints

If you move monitoring components by using node selector constraints, be aware that other constraints can exist for a cluster:

> * Topology spread constraints might exist to control pod scheduling within hierarchical network infrastructure topologies.

Reviewer: we can be more generic, topologies are not necessary "network related": https://kubernetes.io/docs/concepts/scheduling-eviction/topology-spread-constraints/ and https://docs.openshift.com/container-platform/4.12/nodes/scheduling/nodes-scheduler-pod-topology-spread-constraints.html

Reviewer: apart from that, lgtm

Reviewer 3: @machine424 has a point, though I understand we do refer to this use case in another place. Maybe we should relax this a bit? How about:
```suggestion
* Topology spread constraints might exist to control pod scheduling within a cluster's specific topology.
```
Otherwise we could also remove any ambiguity here:
```
* Topology spread constraints might exist.
```

Author: I updated the list item to make it more generic: `Topology spread constraints might be in place to control pod placement.`
